src/circuit/cache.rs

```diff
@@ -0,0 +1,90 @@
+//! Circuit cache.
+//!
+//! # Background
+//!
+//! We use a form of caching to reuse previously constructed fragments of the
+//! circuit. For instance, a
+//! [`stream.integrate()`](`crate::circuit::Stream::integrate`) call creates a
```

Comment:
I think you can be even more general: applying equivalent operators to the same stream produces equivalent streams. In particular, the z inside an integrator is an example.